### DISCUSSION TITLE
Crash in defining StatusBarLabelItem

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#2166] [Q&A] [answered]
I am trying to add a label item to the status bar as shown in https://www.itwinjs.org/learning/ui/abstract/statusbar/

I get the following error when I use the code snippet on this page.  If I use a string instead of ConditionalStringValue for label, crash doesn't happen and the new item is added to the status bar as expected.

For this test, I created a new react app from template (2.19.6) and added a uiProvider (attached) to create a new status bar item.  My relevant code from MyFirstUiProvider is:

  public provideStatusBarItems(stageId: string, stageUsage: string): CommonStatusBarItem[] {
    const iconCondition = new ConditionalStringValue(  
      () => MyFirstUiProvider.labelState ? "icon-home" : "icon-hand-2", 
      ["SampleApp.SET_EXTENSION_UI_VISIBLE"]) 

// From: https://www.itwinjs.org/learning/ui/abstract/statusbar/
      const labelCondition = new ConditionalStringValue(
      () => MyFirstUiProvider.labelState ? "Active" : "Inactive", 
      ["SampleApp.SET_EXTENSION_UI_VISIBLE"])
const widgetItem = AbstractStatusBarItemUtilities.createLabelItem("Sample:StatusBarLabel1", StatusBarSection.Center, 200, iconCondition, labelCondition, undefined);
return [widgetItem];
}

I tried ConditionalStringValue for icon and that works as expected.  

Kindly advise how I can fix this error as  I need to use conditional string for label to achieve what I am trying to show (status of an object in my app)

![StatusBarError](https://user-images.githubusercontent.com/83007711/131381644-850f45fb-01f8-4769-bdcf-f8e4ac29db66.JPG)

Objects are not valid as a React child (found: object with keys {stringGetter, syncEventIds, _value}). If you meant to render a collection of children, use an array instead. in span (created by StatusBarComposer) in div (created by FooterIndicator) in FooterIndicator (created by StatusBarComposer) in div (created by DockedStatusBarItem) in DockedStatusBarItem (created by StatusBarComposer) in DockedStatusbarEntry (created by StatusBarComposer) in div (created by Div) in Div (created by StatusBarCenterSection) in StatusBarCenterSection (created by StatusBarComposer) in div (created by Div) in Div (created by StatusBarSpaceBetween) in StatusBarSpaceBetween (created by StatusBarComposer) in div (created by StatusBarComposer) in StatusBarComposer (created by StatusBar) in div (created by Footer) in div (created by Footer) in div (created by Footer) in Footer (created by Context.Consumer) in StatusBar (created by WidgetPanelsStatusBar) in WidgetPanelsStatusBar (created by WidgetPanelsFrontstageComponent) in WidgetPanelsFrontstageComponent in WidgetContentManager (created by NineZoneProvider) in CursorTypeProvider (created by NineZoneProvider) in DraggedResizeHandleProvider (created by DragProvider) in DraggedPanelSideProvider (created by DragProvider) in DraggedWidgetIdProvider (created by DragProvider) in DragProvider (created by NineZoneProvider) in NineZoneProvider (created by NineZone) in NineZone (created by ActiveFrontstageDefProvider) in div (created by ActiveFrontstageDefProvider) in ActiveFrontstageDefProvider (created by WidgetPanelsFrontstage) in WidgetPanelsFrontstage (created by ConfigurableUiContent) in main (created by ConfigurableUiContent) in ConfigurableUiContent (created by IModelViewer) in FrameworkVersion (created by IModelViewer) in ThemeManagerComponent (created by Connect(ThemeManagerComponent)) in Connect(ThemeManagerComponent) (created by IModelViewer) in IModelViewer in Provider in div in Unknown (created by IModelLoader) in IModelLoader (created by BaseViewer) in ErrorBoundary (created by BaseViewer) in BaseViewer (created by Viewer) in Viewer (at App.tsx:170) in div (at App.tsx:161) in App (at src/index.tsx:13)

[MyFirstUiProvider.zip](https://github.com/imodeljs/imodeljs/files/7078566/MyFirstUiProvider.zip)

**Comment by bsteinbk** [comment, marked as answer]
@ovaisa You found a bug in the label processing. We did not properly process the label when a ConditionalStringValue was used. I will fix this issue in our next major release. Below is the code you sent with a new function component that will provide the same basic functionality. Hope this will get you by until the fix is made.

```
function StatusBarLabelAndIcon({ icon, label }: { icon?: string | ConditionalStringValue, label?: string | ConditionalStringValue }) {
  const [iconSpec, setIconSpec] = React.useState(ConditionalStringValue.getValue(icon));
  const [displayLabel, setDisplayLabel] = React.useState(ConditionalStringValue.getValue(label));

  const syncIdsOfInterest = React.useMemo(() => {
    const ids: string[] = [];
    if (icon instanceof ConditionalStringValue) {
      ids.push(...icon.syncEventIds);
    }
    if (label instanceof ConditionalStringValue) {
      ids.push(...label.syncEventIds);
    }
    return ids;
  }, [icon, label]);

  React.useEffect(() => {
    const handleSyncUiEvent = (args: SyncUiEventArgs) => {
      if (0 === syncIdsOfInterest.length)
        return;

      if (syncIdsOfInterest.some((value: string): boolean => args.eventIds.has(value.toLowerCase()))) {
        if (icon instanceof ConditionalStringValue) {
          icon.refresh();
          setIconSpec(ConditionalStringValue.getValue(icon));
        }

        if (label instanceof ConditionalStringValue) {
          label.refresh();
          setDisplayLabel(ConditionalStringValue.getValue(label));
        }
      }
    };

    SyncUiEventDispatcher.onSyncUiEvent.addListener(handleSyncUiEvent);
    return () => {
      SyncUiEventDispatcher.onSyncUiEvent.removeListener(handleSyncUiEvent);
    };
  }, [icon, label, syncIdsOfInterest]);

  const iconPaddingClass = "nz-icon-padding-right";
  return (
    <FooterIndicator>
      {iconSpec && <Icon iconSpec={iconSpec} />}
      {displayLabel && <span className={iconPaddingClass}>{displayLabel}</span>}
    </FooterIndicator>
  );
}

export class MyFirstUiProvider implements UiItemsProvider {
  public readonly id = MyFirstUiProvider.providerId;
  public static providerId = "MyFirstProviderId";
  public static syncEventId = "my-first-provider:set_extension_ui_visible";

  public static register() {
    if (!UiItemsManager.registeredProviderIds.includes(MyFirstUiProvider.providerId))
      UiItemsManager.register(new MyFirstUiProvider());
  }

  public static toggleLabel() {
    MyFirstUiProvider.labelState = !MyFirstUiProvider.labelState;
    SyncUiEventDispatcher.dispatchSyncUiEvent(MyFirstUiProvider.syncEventId);
  }

  public static labelState = false;

  public provideStatusBarItems(stageId: string, stageUsage: string): CommonStatusBarItem[] {
    if ("ViewsFrontstage" !== stageId || stageUsage !== StageUsage.General)
      return [];

    const iconCondition = new ConditionalStringValue(
      () => MyFirstUiProvider.labelState ? "icon-home" : "icon-hand-2",
      [MyFirstUiProvider.syncEventId]);
    const labelCondition = new ConditionalStringValue(
      () => MyFirstUiProvider.labelState ? "Active" : "Inactive",
      [MyFirstUiProvider.syncEventId]);

    const statusBarItem = StatusBarItemUtilities.createStatusBarItem("MyFirstUiProvider:StatusBarLabel1", StatusBarSection.Center, 200,
      <StatusBarLabelAndIcon icon={iconCondition} label={labelCondition} />);
    return [statusBarItem];
  }

  public provideWidgets(stageId: string, stageUsage: string, location: StagePanelLocation, section?: StagePanelSection): ReadonlyArray<AbstractWidgetProps> {
    const widgets: AbstractWidgetProps[] = [];
    if ("ViewsFrontstage" !== stageId || stageUsage !== StageUsage.General)
      return [];

    if (location === StagePanelLocation.Right && section === StagePanelSection.Start) {
      const helloWidget: AbstractWidgetProps = {
        id: "MyFirstUiProvider.HelloWidget",
        label: "Hello",
        getWidgetContent() {
          return (
            <Button onClick={MyFirstUiProvider.toggleLabel}>Toggle Label</Button>
          );
        },
      };
      widgets.push(helloWidget);
    }

    return widgets;
  }
}
```

**Reply by ovaisa** [reply]
Thank you Bill, the code you provided is working and I am able to do exactly what I was wanting to do.  Much appreciate the extra effort you put in to provide this workaround code - above and beyond!!!  Thanks again!

**Reply by ovaisa** [reply]
I was working to replace the code Bill had provided with the iTwin function for this purpose(```StatusBarItemUtilities.createLabelItem()```).  The bug that was crashing Viewer seems to be fixed and is working properly.  

I found an error that is still crashing the Viewer.   If the value returned by ```ConditionalStringValue()``` for ```icon``` parameter is "", I get a crash shown below.  I can pass "" in the call as an argument and that doesn't seem to cause any issue.  I added a check in my conditional function to prevent this crash but wanted to report this.

Please see the code I originally submitted above.  The following function is changed to show the error.

```
const iconCondition = new ConditionalStringValue(
() => MyFirstUiProvider.labelState ? "" : "",
["SampleApp.SET_EXTENSION_UI_VISIBLE"])
```

```labelCondition``` doesn't cause this crash if the return value is "".

![image](https://user-images.githubusercontent.com/83007711/235998035-bc5372d9-7bcd-4594-98a6-b7cb5744750d.png)
